src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2902,12 +2886,29 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           }
         }
 
-        // if we had missing param types, see if we can undo eta-expansion and recover type info
-        val expectedFunTypeBeforeEtaExpansion =
-          if (paramsMissingType.isEmpty) NoType
-          else typeUnEtaExpanded
+        // If we're typing `(a1: T1, ..., aN: TN) => m(a1,..., aN)`, where some Ti are not fully defined,
+        // type `m` directly (undoing eta-expansion of method m) to determine the argument types.
+        val ptUnrollingEtaExpansion =
```

Comment:
Might be worth a comment to note that this is for the `foo(_, _, _)` form of eta expansion, rather than the `foo _` version (which is encoded as `Typed(..., Function(Nil, EmptyTree)`